src/util/range.h

```diff
@@ -265,6 +265,87 @@ struct concat_iteratort
   second_iteratort second_begin;
 };
 
+/// Zip two ranges to make a range of pairs.
+/// On increment, both iterators are incremented.
+/// Ends when the two ranges reach their ends, raise an exception if one ends
+/// before the other.
```

Comment:
See @smowton's comment below, but I think this should never be an exception in the first place, it should just be an invariant.